rclcpp/topics/minimal_publisher/member_function_with_wait_for_all_acked.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <inttypes.h>
+
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+using namespace std::chrono_literals;
+
+/* This example shows how to use wait_for_all_acked for the publisher */
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher_with_wait_for_all_acked"),
+    count_(0),
+    wait_timeout_(300)
+  {
+    // publisher must set reliable mode
+    publisher_ = this->create_publisher<std_msgs::msg::String>(
+      "topic",
+      rclcpp::QoS(10).reliable());
+
+    // call wait_for_all_acked before shutdown
+    using rclcpp::contexts::get_global_default_context;
+    get_global_default_context()->add_pre_shutdown_callback(
```

Comment:
why "pre shutdown"?
The timer might still be running, so I'm not sure if this is going to wait for messages published after the method was called.
Maybe this should be done in the destructor?
Or after shutdown? Though I'm not sure if we're going to get a "context was shutdown" error in that case.

Reply:
> The timer might still be running.

Yeah. This is a problem. 

> Maybe this should be done in the destructor? 
> Or after shutdown?

wait_for_all_acked() cannot be called after shutdown.  
It is because rcl layer is shutdown. I will cancel timer before calling wait_for_all_acked().

```c++
ontext::shutdown(const std::string & reason)
{
 ...

  // call each pre-shutdown callback
  {
    std::lock_guard<std::mutex> lock{pre_shutdown_callbacks_mutex_};
    for (const auto & callback : pre_shutdown_callbacks_) {
      (*callback)();
    }
  }

  // rcl shutdown
  rcl_ret_t ret = rcl_shutdown(rcl_context_.get());
  if (RCL_RET_OK != ret) {
    rclcpp::exceptions::throw_from_rcl_error(ret);
  }
  // set shutdown reason
  shutdown_reason_ = reason;
  // call each shutdown callback
  {
    std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
    for (const auto & callback : on_shutdown_callbacks_) {
      (*callback)();
    }
  }
 ...
}

Reply:
>  I will cancel timer before calling wait_for_all_acked().

That seems okay.

> It is because rcl layer is shutdown.

I think that methods like `wait_for_all_acked()` should continue working after shutdown.
I don't see the point of making them fail.

Reply:
>>It is because rcl layer is shutdown.
>
> I think that methods like wait_for_all_acked() should continue working after shutdown.
> I don't see the point of making them fail.

From my understanding, once rcl is shutdown, context is invalid. And then if call  wait_for_all_acked, failure is gotten.

Reply:
I understand how it works.
What I mean is that doesn't seem to be useful.

IMO, we don't need much more that notifying executors to stop in practice, I think that calls like `publish()`, `wait_for_all_ack()` or anything else should still be valid after shutdown.

But anyways, the fix you pushed seems fine, and modifying how shutdown works seems out of scope.